docs/design/kuadrant-api-v2.md

```diff
@@ -0,0 +1,178 @@
+
+Currently, as of kuadrant v0.2.0, APIProduct [Group Version Kind](https://book.kubebuilder.io/cronjob-tutorial/gvks.html)
+is: [apiproducts.v1beta1.networking.kuadrant.io](https://github.com/Kuadrant/kuadrant-controller/blob/v0.2.0/doc/architecture.md#api-product-crd)
+
+Taking into account kuadrant is not providing networking level API, but **API Management** level API,
+the proposed GVK is: **apiproducts.v1alpha1.apim.kuadrant.io**.
+
+**Domains**
+
+APIProduct resource will be configured with a list of destination domains or hostnames to which
+traffic is being sent.
+
+It is the responsability of cluster operators and API owners to forward *îngress* traffic for those
+specific domains to the kuadrant gateway. A kuadrant gateway would be a gateway configured by the
+kuadrant control plane, having as input the API Product resource.
+
+For example, cluster operators can deploy required (cloud provider level) networking resources to
+forward traffic to the cluster ingress gateways. Then, the API owners can create (cluster level)
+networking resources, like kubernetes Ingress or Gateway API HttpRoutes, to route that traffic to
+the kuadrant gateways. In this example only *North/South* traffic was referenced. By design,ç
+it could also be *East/west* traffic. By design, there is nothing preventing from having *East/West*
+traffic being protected by kuadrant. This is mainly because kuadrant is not a networking API.
+
+**Protected API specification**
+
+Kuadrant will support the open standard
+[OpenAPI 3.X specification](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.2.md)
+for the API spec.
+
+Each *API Product* can hold multiple OAS docs references. The *API Product* `spec` field includes references
+in the form of URLs and config map references. The actual content may be included in the `status` field.
+
+Kuadrant will extend the OAS doc leveraging
+[OAS custom extensions](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.2.md#specificationExtensions)
+`x-kuadrant-*`, initially for rate limit purposes. In the future, OAS extensions could be leveraged
+for other API management features, for instance usage plans. More about the kuadrant custom
+extensions in the spec reference section.
+
+In case of lacking one OpenAPI doc, kuadrant will provide tooling for easy generation of OpenAPI docs.
+
+Any kuadrant configuration at API path or operation level will be defined as OpenAPI kuadrant extension.
+
+**Backends - Upstreams**
+
+The protected API source, owned by the kuadrant customer, can be specified in two forms:
+* A (base) URL
+* A kubernetes service reference (name and namespace).
+
+When the source is a kubernetes service, kuadrant provides annotations and labels that can be added
+to the service. These annotations and labels can specify, for example, the OAS reference.
+These annotations and labels aim to make the `spec` of the *API Product* resource simpler to specify.
```

Comment:
to encourage committing resources to version control. What if the controller didn't generate the objects in the cluster but instead we had a CLI that generated the resources this could then be used as part of CI/CD flow and also reviewed in code review etc.
In this scenario the controllers job is to just bring the cluster into the state declared by the resources

Reply:
I see your point and definitely something to have in mind. 

Asking the customer to review istio/envoy resources is asking the customer to have knowledge at the networking layer. I do not think it makes sense, but we can discuss further. 

From my current perspective (changing every week or so), kuadrant is a bridge between API management layer and networking layer. Understands the API owner's language (OpenAPI) and configures the networking layer to make it happen. So, asking the customer to review the networking layer manifests breaks that model. It's like telling the customer: "here you have istio, Authorino and RateLimit, go and configure yourself for your own needs".